clusterctl/validation/validate_cluster_api_objects_test.go

```diff
@@ -0,0 +1,531 @@
+			errorMessage: nil,
+			expectErr:    true,
+		},
+		{
+			name:         "Machine has error message",
+			nodeRef:      &testNodeRef,
+			errorReason:  nil,
+			errorMessage: &machineErrorMessage,
+			expectErr:    true,
+		},
+		{
+			name:         "Machine has error reason and message",
+			nodeRef:      &testNodeRef,
+			errorReason:  &machineErrorReason,
+			errorMessage: &machineErrorMessage,
+			expectErr:    true,
+		},
+	}
+	for _, testcase := range testcases {
+		t.Run(testcase.name, func(t *testing.T) {
+			machines := v1alpha1.MachineList{
+				Items: []v1alpha1.Machine{
+					getMachineWithError("test-machine-with-no-error", &testNodeRef, nil, nil),
+					getMachineWithError("test-machine", testcase.nodeRef, testcase.errorReason, testcase.errorMessage),
+				},
+			}
+			var b bytes.Buffer
+			err := validateMachineObjects(&b, &machines, k8sClient)
+			if (testcase.expectErr && err == nil) || (!testcase.expectErr && err != nil) {
+				t.Fatalf("Unexpected returned error. Got: %v, Want Err: %v", err, testcase.expectErr)
+			}
+		})
+	}
+}
+
+func TestValidateMachineObjectWithReferredNode(t *testing.T) {
+	testNodeReadyName := "test-node-ready"
+	testNodeReady := getNodeWithReadyStatus(testNodeReadyName, v1.ConditionTrue)
+	actualTestNodeReady, err := k8sClient.CoreV1().Nodes().Create(&testNodeReady)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer k8sClient.CoreV1().Nodes().Delete(actualTestNodeReady.Name, &meta_v1.DeleteOptions{})
+
+	testNodeNotReadyName := "test-node-not-ready"
+	testNodeNotReady := getNodeWithReadyStatus(testNodeNotReadyName, v1.ConditionFalse)
+	actualTestNodeNotReady, err := k8sClient.CoreV1().Nodes().Create(&testNodeNotReady)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer k8sClient.CoreV1().Nodes().Delete(actualTestNodeNotReady.Name, &meta_v1.DeleteOptions{})
+
+	testNodeNotExistName := "test-node-not-exist"
+
+	var testcases = []struct {
+		name         string
+		nodeRef      v1.ObjectReference
+		expectErr    bool
+	}{
+		{
+			name:         "Machine's ref node is ready",
+			nodeRef:      v1.ObjectReference{Kind: "Node", Name: testNodeReadyName},
+			expectErr:    false,
+		},
+		{
+			name:         "Machine's ref node is not ready",
+			nodeRef:      v1.ObjectReference{Kind: "Node", Name: testNodeNotReadyName},
+			expectErr:    true,
+		},
+		{
+			name:         "Machine's ref node does not exist",
+			nodeRef:      v1.ObjectReference{Kind: "Node", Name: testNodeNotExistName},
+			expectErr:    true,
+		},
+	}
+	for _, testcase := range testcases {
+		t.Run(testcase.name, func(t *testing.T) {
+			machines := v1alpha1.MachineList{
+				Items: []v1alpha1.Machine{
+					getMachineWithError("test-machine", &testcase.nodeRef, nil, nil),
+				},
+			}
+			var b bytes.Buffer
+			err := validateMachineObjects(&b, &machines, k8sClient)
+			if (testcase.expectErr && err == nil) || (!testcase.expectErr && err != nil) {
+				t.Fatalf("Unexpected returned error. Got: %v, Want Err: %v", err, testcase.expectErr)
```

Comment:
The text that gets printed from this line is a bit unclear to me.

In one case, we didn't expect an error but we got one, in that case, this will be printed:
"Unexpected returned error. Got: error message, Want Err: false"
It reads as if the "false" at the end should be an error that we DID expect to get.

In the other case, we did expect an error but didn't get one, in that case it will read:
"Unexpected returned error. Got: nil, Want Err: true"
The first sentence reads as if we got an error, but in fact the error was nil.

It might help to split these cases up so that the text can be different in these 2 cases, what do you think?